### ISSUE TITLE
Proposal: Options for testing JSON responses

### ISSUE THREAD

**Issue by devonestes** [keathley/wallaby#299]
Howdy!

First off, thanks for this wonderful library! It's really nice.

So, today I've been working on refactoring a Phoenix 1.2 app to move to 1.3, including moving to Contexts. While doing that, I wanted to test the top level integration of all my contexts, and to do that I want to write some feature specs that actually go through HTTP (including the router) instead of just using Controller tests as the top level of testing.

Problem is, I'm testing a JSON API. Wallaby seems to be highly optimized for interacting with HTML responses, but less for for JSON responses. It might be nice for those of us who want to test at this level to have some more robust options for doing so. Do you think this is a direction that fits with your vision for this tool?

One thing in particular that I've been looking for for the last 20 minutes or so is some way to access the entire response so I can inspect. I'm sure I could think of other things that would be nice to have, but that's my biggest pain point at the moment.

Thanks again!

**Comment by keathley** [comment]
This is something thats come up before. Right now there isn't a good way for us to capture the response from a json api because its not something that the browsers exposes. We'd have to use something like HTTPoison (which is already a dependency) and then expose an api for testing apis with that. Would that actually solve for your use case @devonestes?

**Comment by devonestes** [comment]
This may be my ignorance of browsers speaking, but I've seen that when you hit an API endpoint in pretty much every browser, it does render the response. For example, https://jsonplaceholder.typicode.com/posts/1 displays in my browser (Chrome).

The reason I thought it might be possible is because I can currently do the following, and it works!

```elixir
assert session |> visit("/posts/1") |> has_text?("Lorem ipsum")
```

So, somewhere in there it's able to see that the session is receiving that JSON response and that it contains the given test. For now that's what I'm using, but it's not the most user friendly API - especially when the assertion fails because I just see `false` and not any indication of what text was in the response.

However, that aside, I think what you propose would definitely work. I don't want to sound too demanding here, but the things that I think would make for a really great user experience for this use case would be:

* The ability to get at the entire response somehow (and log it to the console if you want to see what's being returned). For example, in Capybara I would do `puts JSON.parse(response.body)`.
* Matchers for `has_key?`/`has_keys?` and/or `has_value?`/`has_values?` to test the presence of some response value or just the general shape of the JSON response.

I think with just those two things, combined with the already present `has_text?`, and you'd have a pretty solid tool for _really_ testing a JSON API.

**Comment by keathley** [comment]
I see. I didn't understand that you meant to just visit the api endpoint. A work around might be to do something like this:

```elixir
test "some json", %{session: session} do
  assert session
  |> visit("some.json")
  |> text()
  |> Poison.decode!
  |> get_in(["key", "other_key"]) == "foo"
end
```

I'll do some thinking about what an actual api for this might look like.

**Comment by PragTob** [comment]
:wave: 

If it's all about testing APIs (that's what I understood) then I don't feel like wallaby is the right tool. It's main point is browser testing and for testing an API we don't even need a browser, something that just uses HTTPoison itself and then asserts things on it is much more suitable there I feel. Sort of like controller specs, but going through the whole stack. Or am I missing some crucial part here?

I don't see much of an overlap from the perspective of wallaby there so I see it more as a separate tool.

If it's about JS AJAX calls then I'd wanna say - capturing somehow might be nice, but in the end what's important is the effect it has for the user (somewhat like testing public vs. private methods) - as in, if the call went through somewhere there should be a different UI element which we could check to see if it's there. 

My point of view from my current understanding, might have missed something though :)

**Comment by devonestes** [comment]
Yeah, I can see how this might be outside the intended scope of this project. Makes total sense. In the absence of any other tool, though, I wanted to see if you all felt like this might be a good addition to this library so it could be a sort of "complete solution" for testing your applications in whichever way a real user might interact with them, be it through HTML or JSON. I thought that the tricky part would be the whole part about getting the tests to run in parallel given that they're all integration tests, and using that as the backbone of these API test would be useful.

That workaroud with `text` works for me for now, though, so I'll probably use that until I have the time to play around and try and whip up a quick library to handle this kind of testing.

**Comment by PragTob** [comment]
If we could share the wallaby parallel testing setup, if useful, that would be a cool thing to do. Maybe I'm not too familiar yet though, but I wouldn't hope that it's the biggest hurdle. Personally I'd have nothing against a `wallaby-api` or whatever, but it might be better left to another project. Not sure :)

**Comment by keathley** [comment]
@PragTob I agree that testing apis is a bit outside of wallaby's responsibilities. Not sure if my opinion has changed but I can certainly see the value in being able to re-use Wallaby's concurrent test setup. This is something I've been thinking about for a bit so I'm going to investigate a bit. It would be nice to have concurrent api tests with a really declarative api. Having a separate package like `wallaby-api` might be the way to go.

**Comment by keathley** [comment]
As much as I'd like to do this I don't have time or energy to support it at the moment. If someone else wants to take a crack at it then that would be great. I'm going to close for now though.
